dwh/query_lib.py

```diff
@@ -0,0 +1,278 @@
+              .set('spark.executor.memory', '2g')
+              # See: https://spark.apache.org/docs/latest/security.html
+              .set('spark.authenticate', 'true')
+              )
+      self._spark = SparkSession.builder.config(conf=conf).getOrCreate()
+      # Loading Parquet files and flattening only happens once.
+      self._patient_df = self._spark.read.parquet(self._file_root + '/Patient')
+      self._obs_df = self._spark.read.parquet(self._file_root + '/Observation')
+      # TODO create inspection functions
+      common.custom_log(
+          'Number of Patient resources= {}'.format(self._patient_df.count()))
+      common.custom_log(
+          'Number of Observation resources= {}'.format(self._obs_df.count()))
+      self._flat_obs = _SparkPatientQuery._flatten_obs(self._obs_df)
+      common.custom_log(
+          'Number of flattened obs rows = {}'.format(self._flat_obs.count()))
+    work_df = self._flat_obs.where(self.all_constraints_sql())
+    agg_obs_df = _SparkPatientQuery._aggregate_all_codes_per_patient(work_df)
+    common.custom_log(
+      'Number of aggregated obs= {}'.format(agg_obs_df.count()))
+    self._patient_agg_obs_df = _SparkPatientQuery.join_patients_agg_obs(
+        self._patient_df, agg_obs_df, base_patient_url)
+    common.custom_log('Number of joined patient_agg_obs= {}'.format(
+        self._patient_agg_obs_df.count()))
+    # Spark is supposed to automatically cache DFs after shuffle but it seems
+    # this is not happening!
+    self._patient_agg_obs_df.cache()
+    return self._patient_agg_obs_df.toPandas()
+
+  @staticmethod
+  def _flatten_obs(obs: DataFrame) -> DataFrame:
+    """Creates a flat version of Observation FHIR resources.
+
+    Args:
+      obs: A collection of Observation FHIR resources.
+    Returns:
+      A DataFrame with three columns where `coding` arrays are flattened.
+    """
+    return obs.select(
+        obs.subject.patientId.alias('patientId'),
+        obs.effective.dateTime.alias('dateTime'),
+        obs.value,
+        F.explode(obs.code.coding).alias('coding'))
+
+  @staticmethod
+  def _aggregate_all_codes_per_patient(flat_obs: DataFrame,
+      codes: List[str]=None ) -> DataFrame:
+    """ For each patientId, generates aggregate values for all their observations.
+
+    Args:
+        flat_obs: A collection of flattened Observations.
+        codes: A list of codes for which aggregates are generated or None to
+          indicate all codes.
+    Returns:
+      A DataFrame with one row for each patient and several aggregate columns
+      for each Observation code.
+    """
+    return flat_obs.groupBy([
+        flat_obs.patientId,
+    ]).pivot('coding.code', values=codes).agg(
+        F.count('*').alias('num_obs'),
+        F.min(flat_obs.value.quantity.value).alias('min_value'),
+        F.max(flat_obs.value.quantity.value).alias('max_value'),
+        F.min(flat_obs.dateTime).alias('min_date'),
+        F.max(flat_obs.dateTime).alias('max_date')
```

Comment:
You can add the first and last value here, please see this implementation https://github.com/GoogleCloudPlatform/openmrs-fhir-analytics/pull/177/commits/b6dc16623bc17bd7847d489a0dbe76369f75a90a#diff-1053263adfb29bbd08e6a0723141744b6442c28ca62f9d3aa0813ab7be3cb39eR84-R85

Reply:
thanks @bashir2, let me take a look